Delete user k8s, etc. certificates on re-issue

tcsc opened the pull request:
Prior to this change, user certificates for services like kubernetes were
preserved across a certificate re-issue. This led to issues where elevated
privileges granted by an access request were not applied to the service
certificates as they were not updated during the reissue process.

This patch changes the certificate re-issue process such that:
 * certificates for services (like Kuberenetes) are not preserved
   over a certificate re-issue. It is expected that they will be recreated
   on the first access to the service in question, and
 * the local keystore files for these certificates services are explicitly
   deleted so that the now-invalid cached certificates are not returned
   on keystore queries.

See-Also: #5047

tcsc committed:
Delete user k8s, etc. certificates on re-issue

Prior to this change, user certificates for services like kubernetes were
preserved across a certificate re-issue. This led to issues where elevated
privileges granted by an access request were not applied to the service
certificates as they were not updated during the reissue process.

This patch changes the certificate re-issue process such that:
 * certificates for services (like Kuberenetes) are not preserved
   over a certificate re-issue. It is expected that they will be recreated
   on the first access to the service in question, and
 * the local keystore files for these certificates services are explicitly
   deleted so that the now-invalid cached certificates are not returned
   on keystore queries.

See-Also: #5047

andrejtokarcik commented on lib/client/client.go:
```diff
@@ -186,8 +186,22 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
+	// a new certificate (with the potentially updated roles) on first
+	// access, so we explicitly delete them all.
+	log.Debugf("Deleting all user certs...")
+	if err = localKeyAgent.DeleteUserCerts(params.RouteToCluster, WithAllCerts...); err != nil {
```
Wouldn't this effectively prevent more than a single user/service cert to be stored per cluster at a time? It seems this deletion should occur only if the list of active requests differs from the one used with the stored key.

Also, given the list of active requests should be obtainable from the stored key itself, this logic could be more appropriately moved directly into `LocalKeyAgent.AddKey`.

awly replied:
+1, existing cert deletion should happen only when requesting additional roles, not on every reissue
please move this to `/tool/tsh/`, right after a successful access request approval

tcsc replied:
Moved.

andrejtokarcik commented on lib/client/client.go:
```diff
@@ -186,8 +186,22 @@ func (proxy *ProxyClient) ReissueUserCerts(ctx context.Context, params ReissuePa
+	// access, so we explicitly delete them all.
+	log.Debugf("Deleting all user certs...")
+	if err = localKeyAgent.DeleteUserCerts(params.RouteToCluster, WithAllCerts...); err != nil {
+		log.Errorf("Deleting user certs failed")
```
Assuming writing these logs is really needed here:
```suggestion
		log.WithError(err).Errorf("Deleting user certs failed.")
```

tcsc replied:
Refactored out of existence

andrejtokarcik commented on lib/client/client.go:
```diff
@@ -198,7 +212,13 @@ func (proxy *ProxyClient) reissueUserCerts(ctx context.Context, params ReissuePa
 	if key == nil {
 		var err error
-		key, err = proxy.localAgent().GetKey(params.RouteToCluster, WithAllCerts...)
+		// Note that we are explicitly NOT persisting any existing service
+		// (e.g. DB, Kubernetes, etc) certificates across the re-issue of the
+		// main certificate. This is because the re-issued certificate may
+		// contain new roles, and preserving the old certificates will
+		// only stop those new roles being applied to the service
+		// certificate.
+		key, err = proxy.localAgent().GetKey(params.RouteToCluster)
```
Just to avoid any misunderstanding: This call has nothing to do with *persisting* the key data in the store itself. It's meant just for the convenience of the caller in case they happen to make use of k8s clusters/DBs/apps not mentioned in the reissue request itself.

tcsc replied:
Removed during refactor

awly commented on lib/auth/permissions.go:
```diff
@@ -578,13 +578,13 @@ func contextForBuiltinRole(r BuiltinRole, clusterConfig services.ClusterConfig)
 }
 
-func contextForLocalUser(u LocalUser, identity services.UserGetter, access services.Access) (*Context, error) {
+func contextForLocalUser(u LocalUser, identitySvc services.UserGetter, access services.Access) (*Context, error) {
```
this change seems unrelated to the PR

tcsc replied:
Reverted

tcsc committed:
Address review feedback

andrejtokarcik commented on tool/tsh/tsh.go:
```diff
@@ -2018,20 +2018,58 @@ func reissueWithRequests(cf *CLIConf, tc *client.TeleportClient, reqIDs ...strin
+	// for this oldKey not to exist.
+
+	oldKey, err := tc.LocalAgent().GetKey(profile.Cluster, client.WithSSHCerts{})
+	switch {
+	case trace.IsNotFound(err):
+		oldKey = nil
+	case err != nil:
 		return trace.Wrap(err)
+	default:
+		// carry on
 	}
```
```suggestion
	if err != nil && !trace.IsNotFound(err) {
		return trace.Wrap(err)
	}
```

tcsc replied:
Never mind. I'm drawing the "old" Roles list from the profile now, so the whole thing disappears

tcsc replied:
And the whole think gas been deleted in  any case

andrejtokarcik commented on tool/tsh/access_request_test.go:
```diff
@@ -0,0 +1,149 @@
+		},
+	}
+	keyGen := newKeygen(len(testCases))
+	defer keyGen.Close()
```
Prefer `t.Cleanup` to `defer` in tests.

tcsc replied:
Happy to oblige, but what's the rationale?

andrejtokarcik reviewed (changes requested):
I'm worried the proposed logic is incorrect as it assumes the impact of a role change to be limited to a single cluster whereas I think a role change should discard certs cached for all clusters so that requests against any cluster would trigger a reissue with the new roles.

Further, since SSH certs are now also cached per cluster, they should get discarded with the other service certs as well.

tcsc commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
+		result = append(result, opt)
+	}
+	return result
+}
```
The reissue code expliicitly sets the SSH cert in the key before I get to see it, and deleting it has caused issues downstream. I will attempt to get a "delete everything" approach working, though.

tcsc replied:
"Delete everything" works, it just needs to be inthe middle of `ReissueCertificates()` to work, otherwise the plumbing gets waaaaay too complicated.

tcsc commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
+		return nil
+	}
+
+	// If we get here, we have an old key that matches the new key's index. If
```
Added

tcsc commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
+	// We can force those certs to be re-issued by explicitly deleting them
+	// from the keystore.
+
+	rolesChanged, err := rolesHaveChanged(oldKey, newKey)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	if rolesChanged {
+		err = keyAgent.DeleteUserCerts(newKey.ClusterName, options...)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	return nil
```
Hmm. I can see your point, but I'm not 100% sold on conflating the "error" and "success but nothing more to do" cases.

tcsc replied:
Deleted in change of approach

tcsc committed:
Partial review feedback

tcsc commented:
> Further, since SSH certs are now also cached per cluster, they should get discarded with the other service certs as well.

From wading around in `tsh`, the SSH certs appear to be special; they seem to be the ultimate source of a lot of profile information for `tsh` (e.g. Roles & Active Requests), and a lot of code depends on them just existing at all times. That is not to say that they _shouldn't_ be discarded with the other certs, just that it opens up a lot more complexity than one might think.

I'm not sure of the best way forward with this.

Maybe the correct thing is to stop trying to cobble together a solution out the existing key management primitives, and write something that:
1. Enumerates all the known clusters for the active user behind the given proxy 
2. Deletes the service certs for all of them
3. Generates (somehow!) a new SSH cert for the user/cluster

tcsc committed:
Simplifies everything

Andrew's feedback that it's OK delete the cached certificates without
checking for a role change simplifies everything. This, hopefully last,
patch makes the following changes

 - Deletes all of the role change detection code
 - Adds a parameter to ReissueUserCerts that indicates how existing
   certificates in the cache should be treated (i.e. Kept or Dropped)
 - Adds code to ProxyClient.reissueUserCerts() that takes the
   appropriate cache action.

I've added this to the ProxyClient as the cache cleanup needs to
happen _before_ the new key is added to the key store, otherwise it
becomes very difficult to differentiate between the old certificates
that need purging and new certs that need to be preserved. Given that
the rest of the system currently expects the new certs to be in the key
store before ProxyClient.reissueUserCerts() returns, this looked like
the minimal change required to add the desired behaviour.

tcsc commented on tool/tsh/tsh.go:
```diff
@@ -2018,20 +2018,58 @@ func reissueWithRequests(cf *CLIConf, tc *client.TeleportClient, reqIDs ...strin
+	// the granted roles have changed post-approval. For our purposes here, it's OK
+	// for this oldKey not to exist.
+
+	oldKey, err := tc.LocalAgent().GetKey(profile.Cluster, client.WithSSHCerts{})
```
I've changed the code to take the baseline roleset from the active profile, which is already loaded in this function anyway. This _seems_ like a better place, but please let me know if that's a bad assumption.

tcsc replied:
Deleted in change of approach

awly commented on tool/tsh/tsh.go:
```diff
@@ -2018,20 +2018,58 @@ func reissueWithRequests(cf *CLIConf, tc *client.TeleportClient, reqIDs ...strin
+	//     process, so we don't want to delete that. All of the rest are
+	//     fair game.
+
+	err = deleteUserCertsOnRoleMismatch(oldKey, newKey, newKey.ClusterName,
+		tc.LocalAgent(), client.WithAllCertsExceptSSH...)
```
As @andrejtokarcik mentioned, we have multiple SSH certs now, one for each cluster.
`newKey` will contain the new SSH cert.
How about deleting _all_ certs, including SSH ones, and then writing `newKey` again?
I believe `tc.ActivateKey(cf.Context, newKey)` should do that

tcsc replied:
I'll give it a shot and let you know.

tcsc replied:
Yep, that's the new approach now. Seems to work in latest patch.

awly commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +291,75 @@ func showRequestTable(reqs []types.AccessRequest) error {
+		return false, trace.Wrap(err)
+	}
+
+	oldRoleSet := make(map[string]int)
+	for _, r := range oldRoles {
+		oldRoleSet[r] = 0
+	}
+
+	newRoleSet := make(map[string]int)
+	for _, r := range newRoles {
+		newRoleSet[r] = 0
+	}
```
```suggestion
	oldRoleSet := make(map[string]struct{})
	for _, r := range oldRoles {
		oldRoleSet[r] = struct{}{}
	}

	newRoleSet := make(map[string]struct{})
	for _, r := range newRoles {
		newRoleSet[r] = struct{}{}
	}
```

tcsc replied:
Deleted in favour of unconditionally deleting cert cache, but at least made me look up https://dave.cheney.net/2014/03/25/the-empty-struct

awly commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
+// Note that it's legitimate behaviour for `oldKey` to be nil; this just
+// implies that there are no "old" credentials for us to expire and there
+// is nothing to do.
+func deleteUserCertsOnRoleMismatch(oldKey, newKey *client.Key,
+	clusterName string, keyAgent *client.LocalKeyAgent,
+	options ...client.CertOption) error {
```
@andrejtokarcik why do you think it should be in `LocalKeyAgent`?

andrejtokarcik replied:
Well, from my perspective the whole delete-on-roleset-change logic would ideally occur inside `LocalKeyAgent.AddKey`, similarly to delete-on-private-key-change that's already performed there.

My reasoning is that we should be defensive and not expect explicit role requests to be the only way in which the roleset may be updated (even if it currently is the case). The logic should be made generally available (and even triggered "for free") for new functionality that may change roles or even current use cases that we might've missed.

awly commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +291,75 @@ func showRequestTable(reqs []types.AccessRequest) error {
+	}
+
+	if oldKey.KeyIndex != newKey.KeyIndex {
+		return trace.BadParameter(
+			"old key index %+v does not match new key index %+v",
+			oldKey.KeyIndex, newKey.KeyIndex)
```
nit: don't wrap long lines https://github.com/golang/go/wiki/CodeReviewComments#line-length

tcsc replied:
Deleted anyway

awly commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +291,75 @@ func showRequestTable(reqs []types.AccessRequest) error {
+		return trace.Wrap(err)
+	}
+
+	if rolesChanged {
```
while it's nice to minimize disk IO, doing this check may be unnecessary
this function is only called during an access request flow, so user roles are most likely going to change

WDYT about removing all the roleset checking logic, and deleting the certs unconditionally?

tcsc replied:
That was the initial implementation, so yeah - I have no particular issues there. That would also be a bit more future-proof, should we start granting things _other_ than roles later on.

tcsc replied:
Restored unconditional deletion

awly commented on tool/tsh/access_request_test.go:
```diff
@@ -0,0 +1,161 @@
+	return key, nil
+}
+
+func TestPassingNilToIsNotFoundReturnsFalse(t *testing.T) {
+	require.False(t, trace.IsNotFound(nil))
+}
```
was this for debugging?

tcsc replied:
Just asserting something I wasn't sure of.

tcsc replied:
Deleted

tcsc commented:
@awly's suggestion to go back to unconditionally deleting the certificates without 
checking for a role change simplifies everything. This, hopefully last, patch makes 
the following changes
  
- Deletes all of the role change detection code
- Adds a parameter to `ProxyClient.ReissueUserCerts` that indicates how existing
  certificates in the cache should be treated (i.e. Kept or Dropped)
- Adds handling ProxyClient.ReissueUserCerts() that takes the appropriate cache action.
- Adds handling to ProxyClient.reissueUserCerts() that changes how cached certificates
  are preserved across a re-issue.
    
I've added this to the ProxyClient as the cache cleanup needs to happen _before_ the re-issued `Key` is added to the key store, otherwise it becomes very difficult to differentiate between the old certificates (that need purging) and the new certs (that need to be preserved).

Given that the rest of the system currently expects the new certs to be in the key store before `ProxyClient.ReissueUserCerts()` returns, this looked like the minimal change required to add the desired behaviour.

tcsc commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
+// rolesHaveChanged checks to see if there is a difference in the rolesets
+// applied to two certificates. Handles duplicates and different sort
+// orders in the rolesets.
+func rolesHaveChanged(oldKey, newKey *client.Key) (bool, error) {
```
Deleted in change of approach

tcsc commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
+	// about that if/when this starts to dominate the performance of `tsh`.
+
+	sort.Strings(oldRoles)
+	sort.Strings(newRoles)
```
Deleted in change of approach

tcsc commented on tool/tsh/access_request.go:
```diff
@@ -289,3 +290,97 @@ func showRequestTable(reqs []types.AccessRequest) error {
+		}
+	}
+
+	return false, nil
```
Deleted during change of approach

nklaassen reviewed (approved):
LGTM

andrejtokarcik reviewed (commented):
BTW, does this really work as intended since it doesn't seem to be performing any real deletion of the stored certs anymore?

andrejtokarcik commented on lib/client/client.go:
```diff
@@ -178,27 +178,58 @@ func (p ReissueParams) isMFARequiredRequest(sshLogin string) *proto.IsMFARequire
+
+	// CertCacheKeep indicates that all user certificates (except those
+	// explicitly updated by the re-issue) should be preserved across the
+	// re-issue process
```
```suggestion
	// re-issue process.
```

andrejtokarcik commented on lib/client/client.go:
```diff
@@ -178,27 +178,58 @@ func (p ReissueParams) isMFARequiredRequest(sshLogin string) *proto.IsMFARequire
 	}
 
+	if cachePolicy == CertCacheDrop {
+		proxy.localAgent().DeleteUserCerts("", WithAllCerts...)
```
Missed this, sorry, ignore my question above.

tcsc committed:
Update lib/client/client.go

Co-authored-by: Andrej Tokarčík <andrej@goteleport.com>

russjones reviewed (approved):
Bot.
